### PR TITLE
Review

### PR DESCRIPTION
**Pull request by bl0rch1d**

### CODE REVIEW

File: .ruby-version [hunk 1]
```diff
@@ -0,0 +1 @@
+ext-ruby-2.5.5
```
Comment (yaorlov) [inline .ruby-version]: Why not just `2.5.5`?

File: entities/author.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+class Author
+  attr_accessor :name, :bio
+
+  def initialize(name, bio)
+    @name = name
```
Comment (yaorlov) [inline entities/author.rb:13]: Let's add validation. It should be a String and it shouldn't be empty
Please add validations to other entities as well

File: entities/library.rb [hunk 3]
```diff
@@ -0,0 +1,96 @@
+require 'yaml'
```
Comment (yaorlov) [inline entities/library.rb]: Let's load requires to `autoload.rb` and then require autoload in `main.rb`

File: entities/library.rb [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  def initialize(authors = [], books = [], readers = [], orders = [])
+    @authors = authors
+    @books = books
+    @readers = readers
+    @orders = orders
+  end
+
+  # Library app must have an ability to add each of the entity to itself
+  def add_entity(entity, name)
+    raise ArgumentError unless Config::ENTITIES.include? name.to_sym
```
Comment (yaorlov) [inline entities/library.rb]: Let's use a `case` here:
```
def add_entity(entity)
   case entity
     when Author then @authors << entity
     ...
```

File: entities/library.rb [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+    instance_variable_get("@#{name}s") << entity
+
+    data :save
+  end
+
+  # Library must have the ability to show some custom statistics about the library processes
+  def statistics
```
Comment (yaorlov) [inline entities/library.rb]: Let's move statistics related methods to `Statistics` module

File: entities/library.rb [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+  private
+
+  def for_entities
+    Config::ENTITIES.each do |entity|
+      yield entity
+    end
+  end
+
+  # All the data of the library can be to stored/loaded into some storage (YML,JSON etc)
+  def save_data
```
Comment (yaorlov) [inline entities/library.rb]: Let's move persistence related methods to `Storage/Database` module

File: entities/reader.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
+class Reader
+  attr_accessor :name, :email, :city, :street, :house
+
+  def initialize(name, email, city, street, house)
```
Comment (yaorlov) [inline entities/reader.rb]: I think keyword arguments would be more readable

File: main.rb [hunk 8]
```diff
@@ -0,0 +1,27 @@
+require_relative 'entities/library'
+
+lib = Library.new
+
+# --- Add custom obj for each entity ---
+lib.add_entity Author.new('name', 'bio'), :author
```
Comment (yaorlov) [inline main.rb]: let's use `ffaker/faker` gem  for data generation

File: utils/statistics.rb [hunk 9]
```diff
@@ -0,0 +1,29 @@
+module Statistics
+  def get_top_reader
```
Comment (yaorlov) [inline utils/statistics.rb]: it should accept an argument (default 1)

File: utils/statistics.rb [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    get_top 1, :reader
+  end
+
+  def get_top_books
```
Comment (yaorlov) [inline utils/statistics.rb]: it should accept an argument (default 1)

File: utils/statistics.rb [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    get_top 1, :book
+  end
+
+  def get_readers_of_popular_books
```
Comment (yaorlov) [inline utils/statistics.rb]: it should accept an argument (default 3)

**Review by yaorlov — CHANGES_REQUESTED** [review]
Please run rubocop and fix issues

File: utils/statistics.rb [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+    @orders
+      .select { |order| books.include? order.book }
+      .uniq(&:reader)
+      .size
+  end
+
+  private
+
+  def get_top(num, entity)
+    res = @orders
```
Comment (yaorlov) [inline utils/statistics.rb]: Let's use more descriptive naming

File: utils/statistics.rb [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+          .group_by(&entity)
+          .max_by(num) { |_, x| x.size }
```
Comment (yaorlov) [inline utils/statistics.rb]: `.max_by(num) { |_entity, orders| orders.size }`

**Comment by yaorlov** [comment]
Please add 1 file with stored data (.yml, .json, e.t.c) that is used for data uploading (instead of multiple `#{entity}s.yaml`).
The files that are generated after `save_data` calls should be under gitignore

File: entities/library.rb (hunk 4 above)
Comment (bl0rch1d) [reply entities/library.rb]: А почему case предпочтительней в этом случае? Из-за читаемости?

Comment (yaorlov) [reply entities/library.rb]: из-за читаемости + аргументов меньше

Comment (bl0rch1d) [reply entities/library.rb]: понял, спасибо

File: config.rb [hunk 14]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+module Config
+  DB_PATH = './db/'
+  DB_FILE = 'db.yaml'
+
+  ENTITIES = %i[author book reader order].freeze
```
Comment (yaorlov) [inline config.rb]: this constant is used only in `db.rb`, please move it to that file

File: entities/author.rb [hunk 15]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+# Creates object which represents a book author
+class Author
+  attr_accessor :name, :bio
```
Comment (yaorlov) [inline entities/author.rb]: `attr_reader :name, :bio`?

File: entities/author.rb [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+
+  def initialize(name:, bio:)
+    raise ArgumentError unless valid? name, bio
+
+    @name = name
+    @bio = bio
+  end
+
+  def to_s
+    @name
```
Comment (yaorlov) [inline entities/author.rb]: if you have `attr_reader`/`attr_accessor` you can omit `@`

File: entities/book.rb [hunk 17]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+# Creates object which represents a book
+class Book
+  attr_accessor :title, :author
```
Comment (yaorlov) [inline entities/book.rb]: `attr_reader`?

File: entities/book.rb [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+
+  def initialize(title:, author:)
+    raise ArgumentError unless valid? title, author
+
+    @title = title
+    @author = author
+  end
+
+  def to_s
+    "#{@title}, #{@author}"
```
Comment (yaorlov) [inline entities/book.rb]: "#{title}, #{author}"

File: entities/book.rb [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+  end
+
+  private
+
+  def valid?(title, author)
+    title.instance_of?(String) && !title.empty? && author.instance_of?(Author)
```
Comment (yaorlov) [inline entities/book.rb]: validation logic is repeated in all entities, let's extract validation methods to `Validate` module

File: entities/library.rb [hunk 20]
```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+require_relative '../autoload'
+
+# Creates object which represents Library
```
Comment (yaorlov) [inline entities/library.rb]: please remove explanation comments, the ruby code should be self-explanatory)

File: entities/library.rb [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+# Helps manage the basic needs of Library
+class Library
+  include Statistics
+  include Database
+
+  attr_accessor :authors, :readers, :books, :orders
+
+  def initialize(authors = [], books = [], readers = [], orders = [])
```
Comment (yaorlov) [inline entities/library.rb]: Let's use keyword arguments for better readability:
`def initialize(authors: [], books: [], readers: [], orders: [])`

File: entities/library.rb [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+
+require_relative '../autoload'
+
+# Creates object which represents Library
+# Helps manage the basic needs of Library
+class Library
+  include Statistics
+  include Database
+
+  attr_accessor :authors, :readers, :books, :orders
```
Comment (yaorlov) [inline entities/library.rb]: do we really need `attr_accessor`?

File: entities/library.rb [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+    case entity
+    when Author  then @authors << entity
+    when Book    then @books << entity
+    when Order   then @orders << entity
+    when Reader  then @readers << entity
+    else
+      raise ArgumentError
+    end
+
+    data :save
```
Comment (yaorlov) [inline entities/library.rb]: `add_entity` method should have 1 responsibility, please remove saving data from here

File: entities/library.rb [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+  # Library must have the ability to show some custom statistics about
+  # the library processes
+  def statistics
+    show
+  end
+
+  # Library app must be persisted, thats why we must have the ability
+  # to store/load data from the library.
+  # save/load/generate/delete actions are available
+  def data(action)
```
Comment (yaorlov) [inline entities/library.rb]: this helper method shouldn't be in the library class

File: entities/order.rb [hunk 25]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+# Creates object which represents a store order
+class Order
+  attr_accessor :book, :reader, :date
```
Comment (yaorlov) [inline entities/order.rb]: same comment as for other entities

File: entities/library.rb [hunk 26]
```diff
@@ -0,0 +1,54 @@
...
+    unless private_methods.include? :"#{action}_data"
+      raise ArgumentError, 'Invalid action'
+    end
+
+    unless Pathname.new(Config::DB_PATH).exist?
+      Dir.mkdir(Config::DB_PATH, 0o700)
+    end
+
+    send :"#{action}_data"
```
Comment (yaorlov) [inline entities/library.rb]: better use `public_send`

File: utils/db.rb [hunk 27]
```diff
@@ -0,0 +1,50 @@
...
+
+    @authors  = data[:authors]
+    @books    = data[:books]
+    @orders   = data[:orders]
+    @readers  = data[:readers]
+  end
+
+  # Generating random data using Faker gem
+  def generate_data
+    Config::ENTITIES.each do |entity|
```
Comment (yaorlov) [inline utils/db.rb]: Code with metaprogramming looks shorter, but it's less readable. I would better do smth like this:
```
data = FakeDataGenerator.call # returns a hash {authors: [], books: [], orders: [], readers: []}
authors  = data[:authors]
books    = data[:books]
orders   = data[:orders]
readers  = data[:readers]
```
Also you could extract common code (`authors  = data[:authors]`...) from `load_data` and `generate data` to a separate method.
You don't have to change this, just my thoughts

File: utils/statistics.rb [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+    @orders
+      .select { |order| books.include? order.book }
+      .uniq(&:reader)
+      .size
+  end
+
+  private
+
+  def get_top(quantity, entity_name)
+    res = @orders
```
Comment (yaorlov) [inline utils/statistics.rb]: let's name it `result`

File: entities/library.rb (hunk 20 above)
Comment (bl0rch1d) [reply entities/library.rb]: Я их вписал, потому что линтер ругается на то, что нет top class documentation comment

Comment (bl0rch1d) [reply entities/library.rb]: мне все комментарии убрать?

Comment (yaorlov) [reply entities/library.rb]: для классов можешь оставить (чтоб рубокоп не ругался), убери для методов
а на будущее можешь отключать это правило рубокопа и не писать комментарии на классами

File: entities/library.rb (hunk 26 above)
Comment (bl0rch1d) [reply entities/library.rb]: load_data приватный метод и я не могу использовать public_send

File: utils/db.rb [hunk 29]
```diff
@@ -0,0 +1,61 @@
...
+  def load_data
+    file = DB_PATH + DB_FILE
+    path = Pathname.new(file).exist? ? file : './seed.yaml'
+
+    yaml = File.read(path)
+    data = Psych.safe_load(
+      yaml, [Symbol, Date, Author, Book, Reader, Order], [], true
+    )
+
+    parse data
```
Comment (yaorlov) [inline utils/db.rb]: we are not parsing data here. let's call it `add_to_library` or smth like this

File: utils/fake_data_generator.rb [hunk 30]
```diff
@@ -0,0 +1,61 @@
...
+        books: books,
+        readers: readers,
+        orders: orders
+      }
+    end
+
+    private
+
+    def authors
+      @authors = Array.new(rand(5..LIMIT)) do
```
Comment (yaorlov) [inline utils/fake_data_generator.rb:22]: we don't need to assign array to the instance variable

File: utils/db.rb [hunk 31]
```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+module Database
+  DB_PATH = './db/'
+  DB_FILE = 'db.yaml'
+
+  def data(action)
```
Comment (yaorlov) [inline utils/db.rb]: I don't really like using `send` and raising an exception:
```
unless private_methods.include? :"#{action}_data"
  raise ArgumentError, 'Invalid action'
end
```
Let's remove `data` method and make other methods public (will be used like `library.save_data`)

File: utils/fake_data_generator.rb (hunk 30 above)
Comment (bl0rch1d) [reply utils/fake_data_generator.rb:22]: там дальше есть места, где эти переменные нужны
image

File: utils/validate.rb [hunk 32]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+module Validate
+  def valid?(*props)
```
Comment (yaorlov) [inline utils/validate.rb]: this module should contain helper methods like this:
```
def check_class(instance, klass)
  raise WrongClassError unless instance.instance_of? klass
end
```
Also, let's create `errors.rb` file that will contain error classed for our validations.
```
module Errors
  class WrongClassError < StandardError
    def initialize
      super('Error message')
    end
  end
end
```

File: utils/fake_data_generator.rb (hunk 30 above)
Comment (yaorlov) [reply utils/fake_data_generator.rb:22]: ok

**Review by yaorlov — APPROVED** [review]

**Pull request merged**